lib/ruby_smb/dcerpc/handle.rb

```diff
@@ -0,0 +1,45 @@
+module RubySMB
+  module Dcerpc
+    class Handle
+
+      attr_accessor :pipe
+      attr_accessor :last_msg
+      attr_accessor :msg_type
+      attr_accessor :response
+
+      def initialize(named_pipe)
+        @pipe = named_pipe
+      end
+
+      def bind(endpoint:)
+        ioctl_request(RubySMB::Dcerpc::Bind.new(endpoint: endpoint))
+      end
+
+      def request(opnum:, stub:, options:{})
+        @msg_type = :request
+        ioctl_request(
+            RubySMB::Dcerpc::Request.new(
+                opnum: opnum,
+                stub: stub.new(options).to_binary_s
+            )
+        )
+      end
+
+      def ioctl_request(action, options={})
+        request = @pipe.set_header_fields(RubySMB::SMB2::Packet::IoctlRequest.new(options))
+        request.ctl_code = 0x0011C017
+        request.flags.is_fsctl = 0x00000001
+        request.buffer = action.to_binary_s
+        @last_msg = @pipe.tree.client.send_recv(request)
+        handle_msg(RubySMB::SMB2::Packet::IoctlResponse.read(@last_msg))
+      end
+
+      def handle_msg(msg)
+        if @msg_type == :request
+          dcerpc_response_stub = RubySMB::Dcerpc::Response.read(msg.buffer.to_binary_s).stub
```

Comment:
In my understanding, the output data of a IOCTL Response could be at a different offset in `buffer` ([2.2.32 SMB2 IOCTL Response](https://msdn.microsoft.com/en-us/library/cc246548.aspx)). The `OutputOffset` and `OutputCount` fields can be used to extract the correct output data.
Please have a look at the helper methods I suggest in my comment below (for the lib/ruby_smb/smb2/packet/ioctl_response.rb file)